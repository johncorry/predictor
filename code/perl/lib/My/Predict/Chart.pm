package My::Predict::Chart;
use strict;
use warnings;
use Chart::Gnuplot;

#use Exporter qw(import);
use base 'Exporter';
our @EXPORT_OK = qw(chart_dataset chart_all_datasets);

sub chart_dataset{
   my ($dataset) = @_;

   my $chart = Chart::Gnuplot->new(
      output => "dataset.jpeg",
   );

   my @gnuDataSets;

   my $x_count = 0;
   my @xy = ();
   foreach my $datum (@$dataset) {
      push @xy, [$x_count, $datum];
      $x_count++;
   }
   my $gnuDataSet = Chart::Gnuplot::DataSet->new(
      points => \@xy,
      style => "linespoints"
   );

   push @gnuDataSets, $gnuDataSet;
   $chart->plot2d(@gnuDataSets);
}



sub chart_all_datasets{
   my ($allDatasets) = @_;

   my $chart = Chart::Gnuplot->new(
      output => "datasets.jpeg",
   );

   my @gnuDataSets;

   foreach my $dataset(@$allDatasets) {
      my $x_count = 0;
      my @xy = ();
      foreach my $datum (@$dataset) {
         push @xy, [$x_count, $datum];
         $x_count++;
      }
      my $gnuDataSet = Chart::Gnuplot::DataSet->new(
          points => \@xy,
          style => "linespoints"
      );
      push @gnuDataSets, $gnuDataSet;
   }
   $chart->plot2d(@gnuDataSets);
}

1;
