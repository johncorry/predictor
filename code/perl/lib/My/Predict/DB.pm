package My::Predict::DB;
use strict;
use warnings;

our $Location = "dbi:mysql:PriceData:datastore.c6lgnooprssz.ap-southeast-2.rds.amazonaws.com";
our $User = "jcorry";
our $Pass = "iona22";

1;
