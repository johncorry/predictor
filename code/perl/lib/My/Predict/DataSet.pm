package My::Predict::DataSet;

use strict;
use warnings;

use base 'Exporter';
our @EXPORT_OK = qw(print_dataset print_all_datasets scale_dataset_as_range convert_dataset_to_rate_of_change rolling_Average_Dataset);

sub print_dataset{
   my ($dataset) = @_;
   foreach my $datum (@$dataset) {
      print "$datum\n";
   }
}

sub print_all_datasets{
   my ($allDatasets) = @_;
   foreach my $dataset(@$allDatasets) {
      print "New Dataset\n";
      print_dataset($dataset);
   }
}

sub scale_dataset_as_range{
   my ($dataset) = @_;
   # Apply the same scaling to all values.
   # Assume that the last item is the value to which the others are scaled.
   # The values will also be normalised to be between -1 to 1. 
   
   #print "\nScaling:\n";
   
   #print_dataset($dataset);
   
   my $baseline = @$dataset[-1];
   #print "Baseline = $baseline \n"

   for my $i (0..scalar(@$dataset)) {
      if (defined @$dataset[$i]){
         @$dataset[$i] = @$dataset[$i] - $baseline;
      }
   }

   my $max_datum = 0;

   for my $i (0..scalar(@$dataset)) {
      if (defined @$dataset[$i]){
         if (abs(@$dataset[$i]) > $max_datum){
            $max_datum = abs(@$dataset[$i]);
         }
      }
   }

   if ($max_datum != 0) {
      for my $i (0..scalar(@$dataset)) {
         if (defined @$dataset[$i]){
            @$dataset[$i] = (@$dataset[$i]/$max_datum);
         }
      }
   }

   # FIXME Why are we getting an extra final value printed out after scaling?
   #print_dataset($dataset);
}   

sub convert_dataset_to_rate_of_change{
   my ($dataset) = @_;
   my @datasetTemp = ();

   $datasetTemp[0] = 0; # Don't forget to nullify the first sample.

   for my $i (1..scalar(@$dataset)) {
      if (defined @$dataset[$i]){
         $datasetTemp[$i] = @$dataset[$i] - @$dataset[$i-1];
      }
   }

   for my $i (0..(scalar(@$dataset)-1)) {
      @$dataset[$i] = $datasetTemp[$i];
      #print "$i = " . @$dataset[$i] . "\n";    
   }
   
   shift @$dataset;
}
      
sub rolling_Average_Dataset {

}

1;
