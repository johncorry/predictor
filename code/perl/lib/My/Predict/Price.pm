#!/usr/bin/perl

package Price;

sub new
{
   my $class = shift;
   my $self = {
      _date     => shift,
      _code     => shift,
      _open     => shift,
      _high     => shift,
      _low      => shift,
      _close    => shift,
      _volume   => shift,
      _adjClose => shift,
   };

   bless $self, $class;
   return $self;
}

sub prettyPrint {
   my ( $self ) = @_;

   print "Date     : $self->{_date}\n";
   print "Code     : $self->{_code}\n";
   print "Open     : $self->{_open}\n";
   print "High     : $self->{_high}\n";
   print "Low      : $self->{_low}\n";
   print "Close    : $self->{_close}\n";
   print "Volume   : $self->{_volume}\n";
   print "AdjClose : $self->{_adjClose}\n";
}

sub print {
   my ( $self ) = @_;

   print "$self->{_date},";
   print "$self->{_code},";
   print "$self->{_open},";
   print "$self->{_high},";
   print "$self->{_low},";
   print "$self->{_close},";
   print "$self->{_volume},";
   print "$self->{_adjClose}\n";
}

sub getDate {   
   my ( $self ) = @_;
   return $self->{_date};
}

sub getCode {
   my ( $self ) = @_;
   return $self->{_code};
}
sub getOpen {
   my ( $self ) = @_;
   return $self->{_open};
}

sub getHigh {
   my ( $self ) = @_;
   return $self->{_high};
}

sub getLow {
   my ( $self ) = @_;
   return $self->{_low};
}

sub getClose {
   my ( $self ) = @_;
   return $self->{_close};
}

sub getVolume {
   my ( $self ) = @_;
   return $self->{_volume};
}

sub getAdjClose {
   my ( $self ) = @_;
   return $self->{_adjClose};
}

1;
