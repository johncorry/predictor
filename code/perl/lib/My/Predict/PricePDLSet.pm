#!/usr/bin/perl

package PricePDLSet;
use PDL;

sub new {

   my $class = shift;
   my $priceMap_ref = shift;
   my $dateSeries_ref = shift;

   my @open = ();
   my @high = ();
   my @low = ();
   my @close = ();
   my @volume = ();

   my $prev_open = 0;
   my $prev_high = 0;
   my $prev_low = 0;
   my $prev_close = 0;
   my $prev_volume = 0;


   foreach my $dateKey (@$dateSeries_ref) {

      if (!defined $$priceMap_ref{$dateKey}){
         # Maybe it would better to use the previous price when we don't know.
         #print "NOT DEFINED\n";
         push @open, $prev_open;
         push @high, $prev_high;
         push @low, $prev_low;
         push @close, $prev_close;
         push @volume, $prev_volume;
      }
      else{
         #$$priceMap_ref{$dateKey}->print();
         #print $$priceMap_ref{$dateKey}->getClose() . "\n";
         push @open,   $$priceMap_ref{$dateKey}->getOpen();
         push @high,   $$priceMap_ref{$dateKey}->getHigh();
         push @low,    $$priceMap_ref{$dateKey}->getLow();
         push @close,  $$priceMap_ref{$dateKey}->getClose();
         push @volume, $$priceMap_ref{$dateKey}->getVolume();
         $prev_open = $$priceMap_ref{$dateKey}->getOpen();
         $prev_high = $$priceMap_ref{$dateKey}->getHigh();
         $prev_low = $$priceMap_ref{$dateKey}->getLow();
         $prev_close = $$priceMap_ref{$dateKey}->getClose();
         $prev_volume = $$priceMap_ref{$dateKey}->getVolume();
      }
   }


   my $self = {
     _open_pdl => pdl(@open),
     _low_pdl => pdl(@low),
     _high_pdl => pdl(@high),
     _close_pdl => pdl(@close),
     _volume_pdl => pdl(@volume),
      
   };
   
   my $maxPrice = max($self->{_high_pdl});
 
   if ($maxPrice eq 0){
      $maxPrice = 1;
   }

   $self->{_scaled_open_pdl} = $self->{_open_pdl}/$maxPrice;
   $self->{_scaled_low_pdl} = $self->{_low_pdl}/$maxPrice;
   $self->{_scaled_high_pdl} = $self->{_high_pdl}/$maxPrice;
   $self->{_scaled_close_pdl} = $self->{_close_pdl}/$maxPrice;

   my $maxVolume = max($self->{_volume_pdl});

   if ($maxVolume eq 0){
      $maxVolume = 1;
   }

   $self->{_scaled_volume_pdl} = $self->{_volume_pdl}/$maxVolume;
   
   bless $self, $class;
   return $self;

}

sub getOpenPDL {
   my ( $self ) = @_;
   return $self->{_open_pdl};
}

sub getScaledOpenPDL {
   my ( $self ) = @_;
   return $self->{_scaled_open_pdl};
}

sub getLowPDL {
   my ( $self ) = @_;
   return $self->{_low_pdl};
}

sub getScaledLowPDL {
   my ( $self ) = @_;
   return $self->{_scaled_low_pdl};
}

sub getHighPDL {
   my ( $self ) = @_;
   return $self->{_high_pdl};
}

sub getScaledHighPDL {
   my ( $self ) = @_;
   return $self->{_scaled_high_pdl};
}

sub getClosePDL {
   my ( $self ) = @_;
   return $self->{_close_pdl};
}

sub getScaledClosePDL {
   my ( $self ) = @_;
   return $self->{_scaled_close_pdl};
}

sub getVolumePDL {
   my ( $self ) = @_;
   return $self->{_volume_pdl};
}

sub getScaledVolumePDL {
   my ( $self ) = @_;
   return $self->{_scaled_volume_pdl};
}

1;
