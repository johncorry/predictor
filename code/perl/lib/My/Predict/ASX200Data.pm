#!/usr/bin/perl;

package ASX200Data;

use DBI;
use Time::Piece;
use Time::Seconds;

use My::Predict::DB;
use My::Predict::Price;


my $singleton;

sub new {
   my $class = shift;
   $singleton ||= bless {}, $class;

   #This is what you normally do:
   #bless $self, $class;
   #return $self;

}

sub getStartDate {
   my $self = shift;

   if (exists $self->{_startDate}) {
   }
   else {
      $self->{_startDate} = "2015-06-24";
   }

   return $self->{_startDate};
}

sub getDateSeries {
   my ($self, $code) = @_; 

   my @dateSeries = ();
   my $dateKey = $self->getStartDate();   
   my $datetime = Time::Piece->strptime($dateKey,"%Y-%m-%d");

   #print "\ngetDateSeries\n";

   if (!exists $self->{$code}) {
      $self->getPriceSeries($code);
   }
   
   while ($dateKey le gmtime->strftime("%Y-%m-%d")) {
   #while ($dateKey le "2015-11-03") {


      #if (!defined $$priceMap{$dateKey}){
      if (!defined $self->{$code}{$dateKey}){
         #print "$dateKey NOT DEFINED\n";
      }
      else{
         #print "$dateKey " . "\n";
         push @dateSeries, $dateKey;
      }

      $datetime = $datetime + ONE_DAY;
      $dateKey =  $datetime->strftime("%Y-%m-%d");
   }
 
   return \@dateSeries;
}


sub getPriceSeries {
   my ($self, $code) = @_;

   #print "\ngetPriceSeries\n";

   if (!exists $self->{$code}) {
     $self->{$code} = {};
     #fetch it from the DB
     #Store as a map of Price obj

     # Open DB connection.
     my $dbh = DBI->connect(
       $My::Predict::DB::Location,
       $My::Predict::DB::User,
       $My::Predict::DB::Pass,
       { RaiseError => 0, PrintError => 1 },
     ) or die $DBI::errstr;

     my $sth = $dbh->prepare("SELECT * FROM PriceData.ASX200Price WHERE Code = ?");
     $sth->execute($code);

     while (my @row = $sth->fetchrow_array()) {
       my @words = split / /, $row[0];
       my $date = $words[0];

       my $price = new Price($date,$code,$row[2],$row[3],$row[4],$row[5],$row[6],$row[7]);
       $self->{$code}{$date} = $price;
       #$self->{$code}{$date}->print();
     }
     $sth->finish();
   }

   return $self->{$code};
}

sub getCodes {
   my ($self) = @_;

   if (exists $self->{_codes}){
   }
   else {

      $self->{_codes} = [];

      # Open DB connection.
      my $dbh = DBI->connect(
        $My::Predict::DB::Location,
        $My::Predict::DB::User,
        $My::Predict::DB::Pass,
        { RaiseError => 0, PrintError => 1 },
      ) or die $DBI::errstr;

      my $sth = $dbh->prepare("SELECT Code FROM PriceData.ASX200Company WHERE Sector <> ?");
      $sth->execute("Not Applic");

      while (my @row = $sth->fetchrow_array()) {
         push $self->{_codes}, $row[0];
      }
   } 

   return $self->{_codes};
}

sub getClose {
   my ($self, $code, $date) = @_;

   #print "\ngetClosePrice\n";

   my $price = 0;

   if (!exists $self->{$code}) {
     $self->{$code} = {};
     #fetch it from the DB
     #Store as a map of Price obj

     # Open DB connection.
     my $dbh = DBI->connect(
       $My::Predict::DB::Location,
       $My::Predict::DB::User,
       $My::Predict::DB::Pass,
       { RaiseError => 0, PrintError => 1 },
     ) or die $DBI::errstr;

     my $sth = $dbh->prepare("SELECT Close FROM PriceData.ASX200Price WHERE Code = ? AND Date = ?");
     $sth->execute($code,$date);

     my @row = $sth->fetchrow_array();
     my @words = split / /, $row[0];
     $price = $words[0];
     $sth->finish();
   }

   return $price;
}




1;
