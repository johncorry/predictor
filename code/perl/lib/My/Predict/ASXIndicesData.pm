#!/usr/bin/perl;

package ASXIndicesData;

use DBI;
use Time::Piece;
use Time::Seconds;

use My::Predict::DB;
use My::Predict::Price;


my $singleton;

my %indexLookUp = ('Automobiles & Components','XNJ',
                   'Banks','XFJ',
                   'Capital Goods','XNJ',
                   'Class Pend',"NONE",
                   'Commercial Services & Supplies','XNJ',
                   'Consumer Durables & Apparel','XSJ',
                   'Consumer Services','XSJ',
                   'Diversified Financials','XFJ',
                   'Energy','XEJ',
                   'Food & Staples Retailing','XSJ',
                   'Food, Beverage & Tobacco','XSJ',
                   'Health Care Equipment & Services','XHJ',
                   'Household & Personal Products','XSJ',
                   'Insurance','XFJ',
                   'Materials','XMJ',
                   'Media','XSJ',
                   'Not Applic','NONE',
                   'Pharmaceuticals & Biotechnology','XHJ',
                   'Real Estate','XPJ',
                   'Retailing','XSJ',
                   'Semiconductors & Semiconductor Equipment','XIJ',
                   'Software & Services','XIJ',
                   'Technology Hardware & Equipment','XIJ',
                   'Telecommunication Services','XTJ',
                   'Transportation','XNJ',
                   'Utilities','XUJ'
);

sub new {
   my $class = shift;
   $singleton ||= bless {}, $class;

   #This is what you normally do:
   #bless $self, $class;
   #return $self;

}

sub getStartDate {
   my $self = shift;

   if (exists $self->{_startDate}) {
   }
   else {
      $self->{_startDate} = "2015-03-04";
   }

   return $self->{_startDate};
}

sub getDateSeries {
   my ($self, $code) = @_; 

   my @dateSeries = ();
   my $dateKey = $self->getStartDate();   
   my $datetime = Time::Piece->strptime($dateKey,"%Y-%m-%d");

   #print "\ngetDateSeries\n";

   if (!exists $self->{$code}) {
      $self->getPriceSeries($code);
   }
   
   while ($dateKey ne gmtime->strftime("%Y-%m-%d")) {

      #if (!defined $$priceMap{$dateKey}){
      if (!defined $self->{$code}{$dateKey}){
         #print "$dateKey NOT DEFINED\n";
      }
      else{
         #print "$dateKey " . "\n";
         push @dateSeries, $dateKey;
      }

      $datetime = $datetime + ONE_DAY;
      $dateKey =  $datetime->strftime("%Y-%m-%d");
   }
 
   return \@dateSeries;
}

sub getIndexCode {
   my ($self, $code) = @_;

   #print "\ngetIndexCode\n";

   if (!exists $self->{$code}) {
     $self->{$code} = {};
     #fetch it from the DB
     #Store as a map of Price obj

     # Open DB connection.
     my $dbh = DBI->connect(
       $My::Predict::DB::Location,
       $My::Predict::DB::User,
       $My::Predict::DB::Pass,
       { RaiseError => 0, PrintError => 1 },
     ) or die $DBI::errstr;

     my $sth = $dbh->prepare("SELECT * FROM PriceData.ASX200Company WHERE Code = ?");
     $sth->execute($code);

     my @row = $sth->fetchrow_array();
     $sth->finish();
 
     my @words = split / /, $row[0];
     my $name = $row[1];
     my $sector = $row[2];
     #print "$code $name $sector \n";
     # Need to check this as it might return nothing so we should replace it with "NONE".
     if (!exists $indexLookUp{$sector}){
        $self->{$code} = "NONE";
     }
     else {
        $self->{$code} = $indexLookUp{$sector};
     }
   }

   return $self->{$code};
}

sub getIndexSeries {
   my ($self, $code) = @_;

   #print "\ngetIndexSeries\n";

   if (!exists $self->{$code}) {
     $self->{$code} = {};
     #fetch it from the DB
     #Store as a map of Price obj

     # Open DB connection.
     my $dbh = DBI->connect(
       $My::Predict::DB::Location,
       $My::Predict::DB::User,
       $My::Predict::DB::Pass,
       { RaiseError => 0, PrintError => 1 },
     ) or die $DBI::errstr;

     # 5/1/18 Removed to use ASX200Price table instead.
     # my $sth = $dbh->prepare("SELECT * FROM PriceData.ASXIndicesPrice WHERE Code = ?");
     my $sth = $dbh->prepare("SELECT * FROM PriceData.ASX200Price WHERE Code = ?");
     $sth->execute($code);

     while (my @row = $sth->fetchrow_array()) {
       my @words = split / /, $row[0];
       my $date = $words[0];

       my $price = new Price($date,$code,$row[2],$row[3],$row[4],$row[5],$row[6],$row[7]);
       $self->{$code}{$date} = $price;
       #$self->{$code}{$date}->print();
     }
     $sth->finish();
   }

   return $self->{$code};
}

1;
