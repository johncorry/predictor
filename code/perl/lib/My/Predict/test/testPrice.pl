#!/usr/bin/perl

use strict;
use warnings;

use My::Predict::Price;

my $obj1 = new Price("22-04-1976","JDC",1,2,3,4,5,6);
my $code = $obj1->getCode();
print "$code\n";
$obj1->print();
$obj1->prettyPrint();

