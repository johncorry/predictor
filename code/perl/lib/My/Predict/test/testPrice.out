JDC
22-04-1976,JDC,1,2,3,4,5,6
Date     : 22-04-1976
Code     : JDC
Open     : 1
High     : 2
Low      : 3
Close    : 4
Volume   : 5
AdjClose : 6
